Patch the nearest positions equidistance detection

GiudGiud opened the pull request:
refs #23585 

need this for Javi work asap

GiudGiud committed:
Avoid false positives in detection of nearest position
refs #23585

lindsayad reviewed (changes requested):
How does this work? Code comment would be appreciated

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1684597/) on 9274fec wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25138/site)

This comment will be updated on new commits.

GiudGiud committed:
Address Alex's review
- better cast
- more comments
Update framework/src/positions/Positions.C

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
added some

lindsayad commented:
I still don't understand why this now avoids false positives

GiudGiud commented:
Before we were trying to detect them live in the for loop
so if you look at positions P3 and P4 and they are equidistant to the target (and smaller distance than P1 and P2), we would send a warning. 

but if P5 is actually the closest, then you sent a warning for nothing. P3 and P4 are not the closest positions, who cares that they are equidistant.

now we wait until we have looked at all positions. then send a warning if the closest was also reported as a conflict

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -92,10 +94,18 @@ Positions::getNearestPosition(const Point & target, const bool initial) const
+  if (cast_int<int>(nearest_index) == conflict_index.first ||
+      cast_int<int>(nearest_index) == conflict_index.second)
+  {
+    mooseWarning("Search for nearest position found several matches: " +
```
```suggestion
    mooseWarning("Search for nearest position found at least two matches: " +
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -92,10 +94,18 @@ Positions::getNearestPosition(const Point & target, const bool initial) const
+  }
+
+  // If the closest positions was a conflict, report a warning
+  if (cast_int<int>(nearest_index) == conflict_index.first ||
+      cast_int<int>(nearest_index) == conflict_index.second)
```
```suggestion
  if (cast_int<int>(nearest_index) == conflict_index.second)
```
It should only be possible to match the second right?

GiudGiud replied:
sure

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -92,10 +94,18 @@ Positions::getNearestPosition(const Point & target, const bool initial) const
+      conflict_index = std::make_pair(cast_int<int>(i), cast_int<int>(nearest_index));
+  }
+
+  // If the closest positions was a conflict, report a warning
```
```suggestion
  // If there were multiple "closest" positions, report a warning
```

GiudGiud commented:
looks like not

lindsayad commented:
That wasn't the problem. It was the change in error message

lindsayad commented:
From https://civet.inl.gov/job/1685443/:
```
Search for nearest position found at least two matches
```

lindsayad commented:
Current regex is
```
Search for nearest position found several matches
```

lindsayad commented:
You could revert that change if you want, but I like matching against only what we think we should be matching (in fact I myself might add an assertion to make sure what I think is happening is happening)

GiudGiud commented:
regex changed

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1686821/) on b005126 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>71593a</th>
      <th colspan=4>#25138 b00512</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.61%</td>
      <td>85.61%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89717</td>
      <td>89717</td>
      <td>-</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15078</td>
      <td>15082</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25138/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25138/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25138/coverage/xfem)

</details>



This comment will be updated on new commits.
